### PR TITLE
Pal 392 palisade spring data

### PR DESCRIPTION
**Pull request by dev-958**

### CODE REVIEW

File: palisade-service/src/test/java/uk/gov/gchq/palisade/service/palisade/service/ResultAggregationServiceTest.java [hunk 1]
```diff
@@ -87,7 +82,7 @@ public void setup() {
                 .parent(new DirectoryResource().id("/path/to/")
                         .parent(new DirectoryResource().id("/path/")
                                 .parent(new SystemResource().id("/")))));
-        ConnectionDetail connectionDetail = new SimpleConnectionDetail().uri("data-service");
```
Comment (developer01189998819991197253) [inline palisade-service/src/test/java/uk/gov/gchq/palisade/service/palisade/service/ResultAggregationServiceTest.java:90]: why has this been changed back from hostname to url?

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/ContextConverter.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import java.io.IOException;
+import java.util.Optional;
+
+@Converter
+public class ContextConverter implements AttributeConverter<Context, String> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ContextConverter.class);
+
+    private final ObjectMapper objectMapper;
+
+    public ContextConverter(final ObjectMapper objectMapper) {
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/ContextConverter.java:39]: Please nullcheck this parameter.

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/ContextConverter.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            return this.objectMapper.writeValueAsString(context);
+        } catch (JsonProcessingException e) {
+            LOGGER.error("Could not convert context to json string.", e);
+            return null;
+        }
+    }
+
+    @Override
+    public Context convertToEntityAttribute(final String attribute) {
+        if (Optional.ofNullable(attribute).isPresent()) {
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/ContextConverter.java:58]: Is this the equivalent of if (Objects.isNull(attribute) { try { ....} else { return new Context(); } ? If so, please could it use a normal null check?

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/DataRequestEntity.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        request.setOriginalRequestId(new RequestId().id(this.requestId));
+        return request;
+    }
+
+    public String getRequestId() {
+        return requestId;
+    }
+
+    public void setRequestId(final String requestId) {
+        this.requestId = requestId;
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/DataRequestEntity.java]: Null check

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/DataRequestEntity.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public User getUser() {
+        return user;
+    }
+
+    public void setUser(final User user) {
+        this.user = user;
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/DataRequestEntity.java]: Null check

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/DataRequestEntity.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public Context getContext() {
+        return context;
+    }
+
+    public void setContext(final Context context) {
+        this.context = context;
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/DataRequestEntity.java]: Null check

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/LeafResourceConverter.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            return this.objectMapper.writeValueAsString(leafResource);
+        } catch (JsonProcessingException e) {
+            LOGGER.error("Could not convert leafresource to json string.", e);
+            return null;
+        }
+    }
+
+    @Override
+    public LeafResource convertToEntityAttribute(final String attribute) {
+        if (Optional.ofNullable(attribute).isPresent()) {
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/LeafResourceConverter.java:59]: As above, in ContextConverter

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/DataRequestEntity.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    private Context context;
+
+    @Transient
+    private Map<LeafResource, Rules> leafResourceMap;
+
+    public DataRequestEntity() {
+    }
+
+    public DataRequestEntity(final DataRequestConfig config) {
+        this.requestId = config.getOriginalRequestId().getId();
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/DataRequestEntity.java]: Null check

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/LeafResourceRulesEntity.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Column(name = "rules", columnDefinition = "clob")
+    @Convert(converter = RulesConverter.class)
+    private Rules<?> rules;
+
+    public LeafResourceRulesEntity() {
+    }
+
+    public LeafResourceRulesEntity(final RequestId request, final LeafResource leafResource, final Rules<?> rules) {
+        this.requestId = request.getId();
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/LeafResourceRulesEntity.java]: Null checks on parameters throughout this class needed.

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/RulesConverter.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            return this.objectMapper.writeValueAsString(rules);
+        } catch (JsonProcessingException e) {
+            LOGGER.error("Could not convert rules to json string.", e);
+            return null;
+        }
+    }
+
+    @Override
+    public Rules<?> convertToEntityAttribute(final String attribute) {
+        if (Optional.ofNullable(attribute).isPresent()) {
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/RulesConverter.java:58]: As before

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/UserConverter.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            return this.objectMapper.writeValueAsString(user);
+        } catch (JsonProcessingException e) {
+            LOGGER.error("Could not convert user to json string.", e);
+            return null;
+        }
+    }
+
+    @Override
+    public User convertToEntityAttribute(final String attribute) {
+        if (Optional.ofNullable(attribute).isPresent()) {
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/UserConverter.java:58]: As before.

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/repository/JpaPersistenceLayer.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.stream.Collectors;
+
+import static uk.gov.gchq.palisade.service.palisade.service.PalisadeService.TOKEN_NOT_FOUND_MESSAGE;
+
+public class JpaPersistenceLayer implements PersistenceLayer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(JpaPersistenceLayer.class);
+
+    private DataRequestRepository dataRequestRepository;
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/repository/JpaPersistenceLayer.java]: Make these final

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/repository/JpaPersistenceLayer.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    private LeafResourceRulesRepository leafResourceRulesRepository;
+    private Executor executor;
+
+    public JpaPersistenceLayer(final DataRequestRepository dataRequestRepository, final LeafResourceRulesRepository leafResourceRulesRepository, final Executor executor) {
+        this.dataRequestRepository = dataRequestRepository;
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/repository/JpaPersistenceLayer.java]: Null check these.

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/repository/JpaPersistenceLayer.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        this.leafResourceRulesRepository = leafResourceRulesRepository;
+        this.executor = executor;
+    }
+
+    @Override
+    public DataRequestConfig get(final String requestId) {
+        final List<LeafResourceRulesEntity> leafResourceRules = this.leafResourceRulesRepository.getByRequestId(requestId);
+        final DataRequestEntity dataRequest = this.dataRequestRepository.getByRequestId(requestId);
+
```
Comment (m09526) [inline palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/repository/JpaPersistenceLayer.java:60]: Comment to explain what is being transferred between objects here?

File: palisade-service/src/test/java/uk/gov/gchq/palisade/service/palisade/service/ResultAggregationServiceTest.java (hunk 1 above)
Comment (developer01189998819991197253) [reply palisade-service/src/test/java/uk/gov/gchq/palisade/service/palisade/service/ResultAggregationServiceTest.java:90]: Doesnt really matter what it is, just has to be a string

**Review by developer01189998819991197253 — DISMISSED** [review]

**Review by m09526 — CHANGES_REQUESTED** [review]

File: palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/ContextConverter.java (hunk 3 above)
Comment (dev-958) [reply palisade-service/src/main/java/uk/gov/gchq/palisade/service/palisade/domain/ContextConverter.java:58]: "else" would prevent drop-through on parse exception

**Review by m09526 — APPROVED** [review]

**Pull request merged**